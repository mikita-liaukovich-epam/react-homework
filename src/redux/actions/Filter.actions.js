import { _filter } from '../../models/Filter.DataModel'

const baseSortUrl = '?sortBy=';
const init = { method: 'GET' };

export const fetchSort = (type) => async (dispatch) => {
  const dataUrl = `${Object.keys(_filter).includes(type) ? `${baseSortUrl}/${type}` : ''}`
  const res = await fetch(dataUrl, init)
  const body = await res.json()
  dispatch({ type: _genres.ALL, payload: body })
};