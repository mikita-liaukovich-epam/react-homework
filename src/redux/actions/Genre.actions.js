import {_genres, _genresFriendlyNames } from '../../models/Genres.DataModel'

const baseUrl = '/movies';
const baseSearchUrl = '/movies?searchBy=genres&filter=';
const init = { method: 'GET' };

export const fetchGenre = (type) => async (dispatch) => {
  const res = await fetch(`${Object.keys(_genresFriendlyNames).includes(type) ? baseUrl + '/' + type : '404'}`, init);
  const body = await res.json();
  dispatch({ type: _genres.ALL, payload: body });
};