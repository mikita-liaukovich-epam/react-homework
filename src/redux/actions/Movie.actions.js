import {_genres } from '../../models/Genres.DataModel'

const baseUrl = 'http://localhost:4000/movies';

const fetchMovies = (type, sort) => async (dispatch) => {
  let res;
  const url = `${baseUrl}?${sort ? `sortBy=title&sortOrder=${sort}&` : ''}searchBy=genres&search=${_genres[type] || ''}`;
  res = await fetch(url, { method: 'GET' })
  const body = await res.json()
  dispatch({ type: 'setState', payload: Object.assign(body, { selectedGenre: type }) })
};

const searchMovie = (title) => async (dispatch) => {
  const res = await fetch(`${baseUrl}?searchBy=title&search=${title}`, { method: 'GET' })
  const body = await res.json()
  dispatch({ type: 'setState', payload: body})
}

const addUpdateMovie = (data, isUpdate) => async (dispatch) => {
  const res = await fetch(`${baseUrl}`, {
    method: isUpdate ? 'PUT' : 'POST',
    headers: {
      'Content-type': 'text/plain',
    },
    body: JSON.stringify(data),
  })
  console.log(`Request for movie ${isUpdate ? 'update' : 'creation'} was sent, but swagger is broken and response a Bad Request without a message`, res);
  dispatch({ type: 'setState', payload: { modal: null }})
}

const deleteMovie = (id, selectedGenre) => async (dispatch) => {
  const res = await fetch(`${baseUrl}/${id}`, {
    method: 'DELETE'
  })
  dispatch(fetchMovies(selectedGenre))
}

export {
  addUpdateMovie,
  deleteMovie,
  fetchMovies,
  searchMovie,
}