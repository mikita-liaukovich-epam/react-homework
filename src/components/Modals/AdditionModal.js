import React from "react"
import { useDispatch } from 'react-redux'
import Button from '../Button/Button'
import Form from '../Form/Form'
import ModalBase from './ModalBase'
import { addUpdateMovie } from '../../redux/actions/Movie.actions'

const $ = (id) => document.getElementById(id);

export default function AdditionModal() {
  const dispatch = useDispatch();

  const name = 'addition'

  return <ModalBase modalName={name}>
    <button className="modal-close" onClick={() => dispatch({ type: 'setState', payload: { modal: null }})}>âœ•</button>
    <h2 className="modal-title">Add movie</h2>
    <Form type="title" />
    <Form type="tagline" />
    <Form type="date" />
    <Form type="url" />
    <Form type="genre" />
    <Form type="overview" />
    <Form type="runtime" />
    <Button type="modal" style="filled" onClick={() => {
      const options = {
        title: $('title').value,
        tagline: $('tagline').value,
        vote_average: 0,
        vote_count: 0,
        release_date: $('date').value,
        poster_path: '',
        overview: $('overview').value,
        budget: 0,
        revenue: 0,
        runtime: $('runtime').value,
        genres: [$('genre').value],
      }
      console.log(options)
      dispatch(addUpdateMovie(options))
    }}>Save</Button>
    <Button type="modal" style="outlined" onClick={() => dispatch({ type: 'setState', payload: { modal: null }})}>Reset</Button>
  </ModalBase>
}
