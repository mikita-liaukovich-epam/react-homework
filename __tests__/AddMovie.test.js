import React from 'react'
import AdditionModal from '../src/components/Modals/AdditionModal'
import renderer from 'react-test-renderer'
import ReactTestUtils from 'react-dom/test-utils'
import { Provider } from 'react-redux'
import store from '../src/redux/store'
import * as actions from '../src/redux/actions/Movie.actions'

describe("AdditionModal", () => {
  beforeEach(() => {
    actions.addUpdateMovie = jest.fn();
  })

  it('should render correctly', () => {
    const tree = renderer
      .create(<Provider store={store}><AdditionModal /></Provider>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
})
