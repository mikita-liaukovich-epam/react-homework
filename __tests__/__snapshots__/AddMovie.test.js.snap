// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdditionModal should render correctly 1`] = `
<div
  className="modal addition-modal"
>
  <h1
    className="heading"
  >
    <a
      href="https://netflix.com"
      target="_blank"
    >
      netflix
    </a>
    <span
      className="font_thin"
    >
      <a
        href="javascript:void(0);"
        onClick={[Function]}
      >
        roulette
      </a>
    </span>
  </h1>
  <div
    className="modal-content"
  >
    <button
      className="modal-close"
      onClick={[Function]}
    >
      âœ•
    </button>
    <h2
      className="modal-title"
    >
      Add movie
    </h2>
    <form
      onSubmit={[Function]}
    >
      <div
        className="form-section"
      >
        <h4
          className="form-title"
        >
          Title
        </h4>
        <input
          className="form-input"
          id="title"
          name="title"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Title"
        />
      </div>
      <div
        className="form-section"
      >
        <h4
          className="form-title"
        >
          Tagline
        </h4>
        <input
          className="form-input"
          id="tagline"
          name="tagline"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Tagline"
        />
      </div>
      <div
        className="form-section"
      >
        <h4
          className="form-title"
        >
          Release Date
        </h4>
        <input
          className="form-input form-input_date"
          defaultValue={false}
          id="date"
          name="date"
          onBlur={[Function]}
          onChange={[Function]}
          type="date"
        />
      </div>
      <div
        className="form-section"
      >
        <h4
          className="form-title"
        >
          Movie URL
        </h4>
        <input
          className="form-input"
          id="url"
          name="url"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Movie URL"
        />
      </div>
      <div
        className="form-section"
      />
      <div
        className="form-section"
      >
        <h4
          className="form-title"
        >
          Overview
        </h4>
        <input
          className="form-input"
          id="overview"
          name="overview"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Overview"
        />
      </div>
      <div
        className="form-section"
      >
        <h4
          className="form-title"
        >
          Runtime
        </h4>
        <input
          className="form-input"
          id="runtime"
          name="runtime"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Runtime"
        />
      </div>
      <button
        className="modal-button button_filled"
        type="submit"
      >
        Save
      </button>
      <button
        className="modal-button button_outlined"
        onClick={[Function]}
      >
        Reset
      </button>
    </form>
  </div>
</div>
`;
